packages/wdio-browserstack-service/src/insights-handler.ts

```diff
@@ -0,0 +1,471 @@
+        }
+
+        if (testMetaData && !testMetaData['steps']) {
+            testMetaData['steps'] = []
+        }
+
+        testMetaData['steps']?.push({
+            id: step.id,
+            text: step.text,
+            keyword: step.keyword,
+            started_at: (new Date()).toISOString()
+        })
+
+        this._tests[uniqueId] = testMetaData
+    }
+
+    async afterStep (step: Frameworks.PickleStep, scenario: Pickle, result: Frameworks.PickleResult) {
+        const uniqueId = getUniqueIdentifierForCucumber({ pickle: scenario } as ITestCaseHookParameter)
+        let testMetaData = this._tests[uniqueId]
+        if (!testMetaData) {
+            testMetaData = {
+                steps: []
+            }
+        }
+
+        if (testMetaData && !testMetaData.steps) {
+            testMetaData.steps = [{
+                id: step.id,
+                text: step.text,
+                keyword: step.keyword,
+                finished_at: (new Date()).toISOString(),
+                result: result.passed ? 'PASSED' : 'FAILED',
+                duration: result.duration,
+                failure: result.error ? removeAnsiColors(result.error) : result.error
+            }]
+        } else if (testMetaData){
+            const stepDetails = testMetaData['steps']?.find(item => item.id == step.id)
+            if (stepDetails) {
+                stepDetails.finished_at = (new Date()).toISOString()
+                stepDetails.result = result.passed ? 'PASSED' : 'FAILED'
+                stepDetails.duration = result.duration
+                stepDetails.failure = result.error ? removeAnsiColors(result.error) : result.error
+            }
+        }
+
+        this._tests[uniqueId] = testMetaData
+    }
+
+    async uploadPending () {
+        if (this.requestQueueHandler.pendingUploads <= 0) {
+            return
+        }
+
+        await Promise.race([
+            new Promise(resolve => setTimeout(resolve, 5000)),
+            new Promise(resolve => {
+                if (this.requestQueueHandler.pendingUploads == 0) {
+                    resolve
+                }
+                while (this.requestQueueHandler.pendingUploads > 0) {
+                    // waiting for any pending uploads
+                }
+            }),
+        ])
+    }
+
+    async teardown () {
+        await this.requestQueueHandler.shutdown()
+    }
+
+    /**
+     * misc methods
+     */
+
+    async browserCommand (commandType: string, args: BeforeCommandArgs & AfterCommandArgs, test?: Frameworks.Test | ITestCaseHookParameter) {
+        if (commandType == 'client:beforeCommand') {
+            this._commands[`${args.sessionId}_${args.method}_${args.endpoint}`] = args
+            return
+        }
+
+        if (!test) {
+            return
+        }
+        const identifier = this.getIdentifier(test)
+
+        // log screenshot
+        if (process.env.BS_TESTOPS_ALLOW_SCREENSHOTS == 'true' && isScreenshotCommand(args) && args.result.value) {
```

Comment:
> Also this expect users to set BS_TESTOPS_ALLOW_SCREENSHOTS explicitly to true
Rather than this, we can setting it based on the response from the backend to decide we want to upload screenshots or not.
Updating the logic with `Boolean(process.env.BS_Test Observability_ALLOW_SCREENSHOTS)`